packages/x-date-pickers/src/internals/hooks/useField/useField.ts

```diff
@@ -133,6 +133,125 @@ export const useField = <
     updateSelectedSections();
   });
 
+  const handleInputChange = useEventCallback((event: React.ChangeEvent<HTMLInputElement>) => {
+    if (readOnly || state.selectedSectionIndexes == null) {
+      return;
+    }
+
+    const prevValueStr = state.valueStr;
+    const valueStr = event.target.value;
+
+    let startOfDiffIndex = -1;
+    let endOfDiffIndex = -1;
+    for (let i = 0; i < prevValueStr.length; i += 1) {
+      if (startOfDiffIndex === -1 && prevValueStr[i] !== valueStr[i]) {
+        startOfDiffIndex = i;
+      }
+
+      if (
+        endOfDiffIndex === -1 &&
+        prevValueStr[prevValueStr.length - i - 1] !== valueStr[valueStr.length - i - 1]
+      ) {
+        endOfDiffIndex = i;
+      }
+    }
+
+    const activeSection = state.sections[state.selectedSectionIndexes.start];
+
+    const hasDiffOutsideOfActiveSection =
+      startOfDiffIndex < activeSection.start ||
+      prevValueStr.length - endOfDiffIndex - 1 > activeSection.end;
+
+    if (hasDiffOutsideOfActiveSection) {
+      // TODO: Support if the new date is valid
+      return;
+    }
+
+    // The active section being selected, the browser has replaced its value with the key pressed by the user.
+    const keyPressed = valueStr.slice(
+      activeSection.start,
+      valueStr.length -
+        prevValueStr.length +
+        activeSection.end -
+        (activeSection.separator?.length ?? 0),
+    );
+
+    if (keyPressed.length > 1) {
+      // TODO: Might be able to support it in some scenario
+      return;
+    }
+
+    const isNumericValue = !Number.isNaN(Number(keyPressed));
+
+    if (isNumericValue) {
```

Comment:
The content below was just moved from the `handleInputKeyDown` method, no logic change